src/bootloaders/systemd-class.c

```diff
@@ -174,7 +177,15 @@ static bool sd_class_ensure_dirs(__cbm_unused__ const BootManager *manager)
         return true;
 }
 
-bool sd_class_install_kernel(const BootManager *manager, const Kernel *kernel)
+char *sd_class_get_kernel_dst(const BootManager *manager)
+{
+        (void)manager;
+        return strdup(sd_class_config.kernel_dir);
+}
+
+bool sd_class_install_kernel_impl(const BootManager *manager, const Kernel *kernel,
+                                  char *(*get_kernel_dst)(const BootManager *),
+                                  bool (*ensure_layout)(const BootManager *))
```

Comment:
Seems odd to pass function pointers around like this, and not the actual bootloader struct or similar. Makes the API look kinda janky.

Reply:
Alternatively having sane typedefs for the function pointers would clean it up.

Reply:
I, personally, do not see anything clunky about passing a pointer to a function (because it's C). The logic requires this so-called class to call different implementations depending on the context. I see the problem with the design (which I agree exists) differently: to me the root cause of these pirouettes is the fact that `BootManager` is completely opaque. The very best solution to this is to expose the structure of `BootManager`, at least partially, to be able to access (read-only) the context information and get rid of these completely unnecessary `_impl` functions. However, the purpose of this PR is to make a functional implementation without burying in design discussions.

I'll make the change to pass `BootLoader` to `_impl` functions, but, again, it only masks the design issue (so would `typedef`ing the functions).